Create parsing for texvc.sty using texutil.js

notjagan opened the pull request:
It currently puts the parsed array into other_literals3.

Commit:
Create parsing for texvc.sty using texutil.js

physikerwelt commented:
@notjagan please fix the failing tests

physikerwelt commented:
@HowardCohl @notjagan what is the progess here?

HowardCohl commented:
@physikerwelt It turns out that this was a pull request made by Claude, not @notjagan . Also, it has to do with the texvcjs repo, @notjagan is working on the texer repo.

notjagan committed:
Change texvc.sty parsing and fix failing tests.

notjagan committed:
Remove console.log for testing.

notjagan commented:
While I was getting the code to pass the tests, I noticed two possible inconsistencies between texvc.sty and the preexisting test cases. The first is that the test cases look for `\or` to be replaced with `\lor`, whereas texvc.sty defines the replacement as `\orMediaWiki` for `\lor`. Since the texvc.sty definition only appears once and the test cases had `\or` twice, I changed texvc.sty so it defined the replacement for `\or`. Additionally, in texvc.sty, the replacement for `\varcopper` was defined as `\mbox{\\coppa}`, but the double backslash didn't make sense and wasn't even recognized by the pegJS, so I removed the second backslash from texvc.sty. If either of these changes were incorrect, please notify me and I can change them back.

notjagan committed:
Remove test output file.

physikerwelt commented on lib/npm-debug.log:
```diff
@@ -0,0 +1,18 @@
+0 info it worked if it ends with ok
```
Remove the log file, please.

physikerwelt commented on lib/texutil.js:
```diff
@@ -518,7 +634,7 @@ module.exports.other_literals2 = arr2set([
 ]);
 
-module.exports.other_literals3 = obj2map({
+/*module.exports.other_literals3 = obj2map({
```
Remove the old coede.

physikerwelt commented on lib/texutil.js:
```diff
@@ -1108,4 +1224,4 @@ module.exports.teubner_required = arr2set([
 module.exports.mhchem_required = arr2set([
     "\\ce"
-]);
+]);
```
the file should end with a newline

physikerwelt commented on lib/texvc.sty:
```diff
@@ -0,0 +1,117 @@
+\renewcommand{\O}{\emptyset}
+\newcommand{\omicron}{\mathrm{o}}
+\newcommand{\Omicron}{\mathrm{O}}
+\newcommand{\or}{\lor}
```
The problem is that \or does not compile, with TeX. Thus, we need to develop a workaround for that.

HowardCohl replied:
I have no idea what @physikerwelt means? If it does not compile with TeX, why is it there?

physikerwelt replied:
@HowardCohl @notjagan I suggest the following: We create a separate issue for the \or issue and keep the file as it is of now.

physikerwelt commented on lib/texutil.js:
```diff
@@ -42,11 +42,128 @@ module.exports.latex_function_names = arr2set([
     var fs = require('fs');
     var path = require('path');
+    module.exports.other_literals3 = {};
+    parseSty(fs.readFileSync(path.join(__dirname, "texvc.sty"), 'utf8'), module.exports.other_literals3);
```
This file is already huge. Maybe you can put the parseSty functionality to a seperate file.

physikerwelt replied:
Why not make parseSty return the array. i.e
`module.exports.other_literals3 =parseSty (...)`

physikerwelt commented on lib/texutil.js:
```diff
@@ -42,11 +42,128 @@ module.exports.latex_function_names = arr2set([
+
 module.exports.paren = arr2set(["\\sinh"]);
+module.exports.other_literals3 = obj2map({});
 try{ //try catch to make sure program still runs without csv file
```
I don't think this is a good idea. If something is broken the program should fail and no silently ignore the error.

physikerwelt commented on lib/texutil.js:
```diff
@@ -42,11 +42,128 @@ module.exports.latex_function_names = arr2set([
+
+function parseSty(inp, cmds) {
+    var starts = ["\\newcommand", "\\renewcommand", "\\DeclareRobustCommand", "\\defSpecFun"];
+    inp = inp.split("\n").map(function(line) {return line.match(/([^\\%]|\\.|^)+/g)[0];}).join("\n");
```
Use newlines for the function body.

physikerwelt replied:
move up the definition of lines from below, and reuse it.

physikerwelt replied:
I would recommend to either use more comments or use descriptime methodnames. @HowardCohl and me should be able to understand what each line of the program does.

physikerwelt commented on lib/texutil.js:
```diff
@@ -42,11 +42,128 @@ module.exports.latex_function_names = arr2set([
+    inp = inp.replace(/\n+/g, "\n").replace(/^\n|\n^/g, "");
+    var lines = inp.split("\n");
+    var lineno = 0;
+    while (lineno < lines.length) {
```
can't you map a function to lines?

physikerwelt replied:
I would also make a new function like parseLine so that there is no deep level of nesting

physikerwelt commented on lib/texutil.js:
```diff
@@ -42,11 +42,128 @@ module.exports.latex_function_names = arr2set([
+    }
+}
+
+function bracketBalance(string) {
```
The idea of unit tests is that each small functional unit has the corresponding unit tests.
Maybe you can a few tests of each of the new method you introduce.

notjagan committed:
Restructure parseSty into a separate file and add testing for it.

notjagan committed:
Fix BitHound issues.

notjagan committed:
Reduce function complexity.

notjagan committed:
Fix out-of-scope references.

notjagan committed:
Unify 'if' spacing.

notjagan committed:
Further reduce complexity (most likely temporary).

physikerwelt commented on lib/parseSty.js:
```diff
@@ -6,33 +6,9 @@ var starts = ["\\newcommand",
-    }
-    sections = sections.slice(0, sections.length - 1);
-    return sections;
+    return string.match(/\{[^\{\}]*?(([^\}]*)\}+)*?[^\{\}]*?\}|\[[^\[\]]*?(([^\]]*)\]+)*?[^\[\]]*?\]/g).map(function(s) {
```
In Germany one calls that "to reinvent the wheel".
How about using a well tested library for that like
https://www.npmjs.com/package/balanced-match
?

physikerwelt commented on lib/index.js:
```diff
@@ -18,6 +18,25 @@ module.exports.SyntaxError = Parser.SyntaxError;
 var astutil = require('./astutil');
 var contains_func = module.exports.contains_func = astutil.contains_func;
 
+function handleTexError(e, options) {
```
nice. You should also do a pull requst for this change in the [upstream repo](https://github.com/wikimedia/texvcjs).

notjagan replied:
Should I do the pull request now or wait until this pull request is approved?

physikerwelt replied:
Those two tasks are independent of each other, so you can go ahead and make the pull request. However, this is not high priority.

physikerwelt commented on lib/index.js:
```diff
@@ -28,15 +47,15 @@ var check = module.exports.check = function(input, options) {
+        options.usemathrm = false;
+        options.usmhchem = false;
+        options.semanticlatex = false;
+    }
```
:+1:

physikerwelt commented on lib/index.js:
```diff
@@ -28,15 +47,15 @@ var check = module.exports.check = function(input, options) {
-        }
         // allow user to pass a parsed AST as input, as well as a string
-        if (typeof(input)==='string') {
+        if (typeof(input) === 'string') {
```
:+1:

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,101 @@
+"use strict";
+
+var starts = ["\\newcommand",
```
This variable name is not self explaining. Either rename it, or add a comment what this array is supposed to do.

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,101 @@
+              "\\defSpecFun"];
+
+function findBracketSections(string) {
+    return string.match(/\{[^\{\}]*?(([^\}]*)\}+)*?[^\{\}]*?\}|\[[^\[\]]*?(([^\]]*)\]+)*?[^\[\]]*?\]/g).map(function(s) {
```
this regexp is hard to read, and would need at least a comment. Also take into account my earlier comment about using a library.

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,101 @@
+              "\\DeclareRobustCommand",
+              "\\defSpecFun"];
+
+function findBracketSections(string) {
```
It would also be nice to have a test function for findBracketSections
In addition to prevent bugs, it also helps other to understand what the function does.

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,101 @@
+    });
+}
+
+function findFirst(string, subs, start) {
```
Please create a few test cases for each method.

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,101 @@
+function parseNewCommand(definition, command, cmds) {
+    definition = definition.substring(command.length);
+    var sections = findBracketSections(definition);
+    if (sections.length === 0) {
```
are you aware of http://www.w3schools.com/js/js_switch.asp

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,101 @@
+        if (sections[0][1] !== "[") {
+            cmds[sections[0][0]] = sections[1][0];
+        }
+    }
```
I suggest the program sould error out of sections.length>2.

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,101 @@
+    }
+}
+
+function bracketBalance(string) {
```
Please create a few test cases for each method.

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,101 @@
+    if (typeof cmds === "undefined") {
+        cmds = {};
+    }
+    inp = inp.replace(/\n+/g, "\n").replace(/^\n|\n^/g, "");
```
does this remove empty lines?
You should state that you do not support windows line ends, or add it.
https://github.com/yoshuawuyts/newline-remove/blob/master/index.js#L17
Shows support for different line end encoding.

physikerwelt commented on lib/texutil.js:
```diff
@@ -43,10 +45,19 @@ module.exports.latex_function_names = arr2set([
+catch (e) {
+    if (e.code !== "ENOENT") {
+        throw e;
+    }
```
The file is in the github repo. So it it's OK to error out, the file was not found.

physikerwelt commented on lib/texutil.js:
```diff
@@ -59,8 +70,12 @@ try{ //try catch to make sure program still runs without csv file
-} catch (e) {}
-
+}
+catch (e) {
```
I'm still not a fan of this catch expression.
Why not check with the fs library to see if the file exists?
Moreover, do we still need the csv parser?
@HowardCohl Does it make sense to maintain two ways to define new macros?

**edited by @HowardCohl

HowardCohl replied:
@physikerwelt  Do you mean like using \newcommand and \defSpecFun ?

HowardCohl replied:
@physikerwelt @notjagan  Yes, we are not going to use csv files anymore. We should just remove all references to that, right?

physikerwelt replied:
Yes.

physikerwelt commented on lib/texutil.js:
```diff
@@ -518,82 +533,6 @@ module.exports.other_literals2 = arr2set([
     "\\varstigma"
 ]);
 
-module.exports.other_literals3 = obj2map({
```
:+1:

physikerwelt commented on test/parseSty.js:
```diff
@@ -0,0 +1,26 @@
+            assert.deepEqual(parseSty(tc.input), tc.output);
+        });
+    });
+});
```
:+1:

physikerwelt commented on test/parseSty.sty:
```diff
@@ -0,0 +1,8 @@
+\newcommand{\testC}{\TeC}
+\newcommand{\testD}{\mbox{\TeD}}
+
+\endinput
```
:+1:

physikerwelt commented:
I went through the pull request again. It looks much better now. However, there are still some suggestions I have.
Especially the  lib/parseSty.js  is still hard to understand. It would be good to write the code in a way that you, other students and me are able to improve it in the future afer not having looked at the code for a while.

notjagan committed:
Remove csv parsing code.

notjagan committed:
Restructure parseSty.js.

notjagan committed:
Add test for findBracketSections in parseSty.js.

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,118 @@
+                           "\\defSpecFun"];
+
+var findBracketSections = module.exports.findBracketSections = function(string, sections) {
+    if (typeof sections === "undefined") {
```
I suggest to use `if ( !sections )` instead

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,118 @@
+    }
+    var bracket = balanced("[", "]", string);
+    var brace = balanced("{", "}", string);
+    if (typeof brace !== "undefined" && (typeof brace !== undefined || brace.start < bracket.start)) {
```
what if bracket is undefined? (You can use !brace as above).

physikerwelt replied:
If I understand the code right the logic is, if either of bracket or brace is undefined use the other. If neither is undefined use the first one, correct?

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,118 @@
+    return sections;
+};
+
+function findFirst(string, subs, start) {
```
method needs test.

notjagan committed:
Fix typo in parseSty.js.

physikerwelt commented:
@notjagan please create tests for each method in parseSty.js
Moreover, try to make your code more readable. To do that there are two options. Either simplify the code or add comments.

notjagan commented:
@physikerwelt I am adverse to using `if (!sections)` or anything similar because this just checks if `sections` is falsy. Although this is not really a concern with arrays such as `sections` in my code, other types like integers do encounter issues here, so for consistencies sake I think I should stick to using `if (typeof sections === "undefined")` or switch to `if (sections == null)`, which is the standard method of checking if a variable is `null` or `undefined`. The reason I was not already using `if (sections == null)` is because I wasn't sure if jslint would give me issues because it doesn't use `===`.

physikerwelt commented:
@notjagan I know. But using ! is more common and easier to read. I started with the undefined method as well but thereafter I was advised by more experienced developers that ! should be used instead.

notjagan committed:
Change undefined checks.

notjagan committed:
Add comment in parseSty.js.

notjagan committed:
Fix condition in parseSty.js.

notjagan committed:
Add support for Windows line breaks (\r\n).

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,133 @@
+                           "\\DeclareRobustCommand",
+                           "\\defSpecFun"];
+
+// Returns a two dimensional array, where each element contains the body of a bracket section and the type of bracket
```
Use multiline comments here as in https://github.com/wikimedia/service-template-node/blob/master/routes/ex.js#L44-L49

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,133 @@
+        sections.push([bracket.body, "["]);
+        return findBracketSections(bracket.post, sections);
+    }
+    return sections;
```
other than my first coomment this method is OK.

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,133 @@
+// Finds the first occurrence in the provided string of the earliest occurring item in the given array, starting from
+// a specified index. Returns both the index of occurrence and the relevant item of subs (returns -1 and the first item
+// if none of the provided substrings can be found).
+function findFirst(string, subs, start) {
```
Add test cases for this method, please!

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,133 @@
+}
+
+// Parses the first command declaration of the lines provided.
+function parseLine(lines, cmds) {
```
Add test cases for this method, please!

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,133 @@
+
+// Finds the net level of brackets, braces, parentheses.
+// ([]) -> 0, [[] -> 1, }} -> -2, etc.
+function bracketBalance(string) {
```
Add test cases for this method, please!

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,133 @@
+}
+
+// Parses \newcommand .. ..
+function parseNewCommand(definition, command, cmds) {
```
Add test cases for this method, please!

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,133 @@
+    return sections;
+};
+
+// Finds the first occurrence in the provided string of the earliest occurring item in the given array, starting from
```
I'd call the _string_  _search_ or something. 
Indicate that subs is an array 
http://usejsdoc.org/about-block-inline-tags.html

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,133 @@
+    return [min, first];
+}
+
+// Parses \newcommand .. ..
```
Please describe the method paramters as above.

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,133 @@
+    switch (sections.length) {
+        // \newcommand\command1\command2
+        case 0:
+            var commands = definition.match(/\\[a-zA-Z]+/g);
```
The pattern originates from the restriction of latexmacronames. Can you add a reference to this fact here?

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,133 @@
+
+// Parses \newcommand .. ..
+function parseNewCommand(definition, command, cmds) {
+    definition = definition.substring(command.length);
```
Withing documentation of the parameters I do not understand this line:-(
So I can not continue to review this method.

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,133 @@
+    if (!cmds) {
+        cmds = {};
+    }
+    inp = inp.replace(/(\r\n)+|\n+/g, "\n").replace(/^\n|\n^/g, "");
```
`(\r?\n)`

notjagan committed:
Improve comments on parseSty.js.

notjagan committed:
Add tests and improve comment style for parseSty.js

notjagan commented:
I wrote tests for a majority of the methods within `parseSty.js`. However, there are two methods (`parseLine` and `parseNewCommand`) I haven't written test cases for because they are heavily (and exclusively) used by another parent function (`parseSty`), and the parent function already has test cases written for it. As such, they are already tested by the test cases for the parent function. Additionally, the functions do not return anything of use and only do in-place modification of passed arrays, so test cases in the current style would not be fitting for the checks required to test them independently.

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,177 @@
+            cmds[commands[0]] = commands[1];
+            break;
+        case 1:
+            var cmd = definition.replace(new RegExp(sections[0][0].replace(/\\/g, "\\\\"), 'g'), "").match(/\\[a-zA-Z]+/)[0];
```
It's still not clear to me, what this line does. @HowardCohl do you understand that?

physikerwelt replied:
Maybe it would make sense to introduce a a variable for sections[0] ?
Would that be \command1,'{' int the example below.

HowardCohl replied:
@physikerwelt @notjagan says he fixed this in the latest pull request

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,177 @@
+            break;
+        case 1:
+            var cmd = definition.replace(new RegExp(sections[0][0].replace(/\\/g, "\\\\"), 'g'), "").match(/\\[a-zA-Z]+/)[0];
+            // \newcommand\command1{\replacement}
```
is 1 the argument to command? or did you intend to write commandI?

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,177 @@
+            if (definition.indexOf(cmd) === 0) {
+                cmds[cmd] = sections[0][0];
+            }
+            // \newcommand{\replacement}\command2
```
commandII?

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,177 @@
+    if (!cmds) {
+        cmds = {};
+    }
+    inp = inp.replace(/(\r?\n)+/g, "\n").replace(/^\n|\n^/g, "");
```
What does this line do?

physikerwelt commented on lib/parseSty.js:
```diff
@@ -0,0 +1,177 @@
+        cmds = {};
+    }
+    inp = inp.replace(/(\r?\n)+/g, "\n").replace(/^\n|\n^/g, "");
+    var lines = inp.split("\n");
```
Is it possible that a single command spans multiple lines?

notjagan replied:
What do you mean?

physikerwelt commented:
Well done! The comments help to understand your code. However, there are some things that still are not clear to me.

physikerwelt commented:
@notjagan each function should have at least one test case. But you can create different ways of testing your methods.

notjagan committed:
Add more documentation and tests and improve style

notjagan committed:
Add trailing newline to test file for parseSty.js

physikerwelt commented on test/parseSty.js:
```diff
@@ -46,10 +51,68 @@ var testCases = [
+            parseSty.newCommand("\\newcommand\\test\\test\\test", "\\newcommand", {});
+        }, function(err) { return err instanceof Error; });
+    });
+});
```
newline at the end of the file please

notjagan committed:
Add trailing newline (again?)

physikerwelt commented on .gitignore:
```diff
@@ -50,3 +50,6 @@ crashlytics-build.properties
 lib/optionalFunctions.csv
+
+# Node log
+lib/npm-debug.log
```
add a newline at the end of the file, please.

physikerwelt commented:
only one more newline and you are done ;-)

notjagan committed:
Another newline
